### PR TITLE
HOTT-940 Fix and extend the Search for quotas

### PR DESCRIPTION
**Pull request by alexdesi**
### What?

This PR Fixes and extends the Search for quotas

The search is smarter now, it supports:
- Chapter
- Heading
- Commodity
using the ancestries,
otherwise - in case the term does not look one of those - it uses a ilike search.

### Why?
I am doing this because:

To allow users to find easily quotas.

### Jira link
https://transformuk.atlassian.net/browse/HOTT-940

### CODE REVIEW

File: lib/api_constraints.rb [hunk 1]
```diff
@@ -1,19 +1,14 @@
 class ApiConstraints
-  def initialize(version:)
-    @version = version
+  def initialize(options)
```
Comment (jebw) [inline lib/api_constraints.rb:2]: This changes the method signature from always requiring a version, to the version being optional - I'm not sure if thats a good idea - it could instead be 

```
def initialize(version:, default: nil)
  @version = version
  @default = default
end
```

In general I think changing the API behaviour is a separate thing from this ticket should be in its own PR and reviewed independently

File: spec/services/quota_search_service_spec.rb [hunk 2]
```diff
@@ -1,3 +1,4 @@
+# rubocop:disable RSpec/MultipleMemoizedHelpers
```
Comment (jebw) [inline spec/services/quota_search_service_spec.rb:1]: I appreciate this isn't your code but lots of memoized helpers is a bit of a code smell - I think either we should be deciding to ignore this for now (have have lint fail - allowing the next developer to revisit it) or fixing the issue.

Is it worth adding either extending the factories to allow easily building the desired scenario, or just adding a `shared_context` block to build up the desired data?

**Review by jebw — CHANGES_REQUESTED** [review]
Looks great but I think the Api constraints changes should be spun out into their own PR. 

Does the calling of `Commodity#ancestors` end up with an N+1 ?

**Comment by jebw** [comment]
FYI: If you rebase onto master, that'll fix the CI errors on deploy

File: spec/services/quota_search_service_spec.rb (hunk 2 above)
Comment (alexdesi) [reply spec/services/quota_search_service_spec.rb:1]: yeah, all those lets make the code a bit smelly, I'll see how we can reduce those.

File: app/services/find_ancestors_service.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  def goods_nomenclature
+    goods_nomenclature_class.by_code(goods_nomenclature_item_id).take
+  end
+
+  def goods_nomenclature_class
+    klass_name = GoodsNomenclature.class_determinator
+                                  .call(goods_nomenclature_item_id: goods_nomenclature_item_id)
+
+    StringToClass.convert(klass_name)
```
Comment (willfish) [inline app/services/find_ancestors_service.rb:22]: Does constantize not work?

File: lib/string_to_class.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+module StringToClass
+  ALLOWED_CLASSES = [Chapter, Heading, Commodity].freeze
```
Comment (willfish) [inline lib/string_to_class.rb:2]: This is wayyyyy to defensive. We control the factory definition and if a person has access to commit changes to that we've got significant and various other problems

File: lib/api_constraints.rb (hunk 1 above)
Comment (willfish) [reply lib/api_constraints.rb:2]: I've already made this change @alexdesi (see master)

Comment (willfish) [reply lib/api_constraints.rb:2]: We default from the value in the environment https://github.com/trade-tariff/trade-tariff-backend/pull/219

File: spec/services/find_ancestors_service_spec.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
+describe FindAncestorsService do
+  subject(:find_ancestors) { described_class.new(goods_nomenclature_item_id).call }
```
Comment (willfish) [inline spec/services/find_ancestors_service_spec.rb:2]:
```suggestion
  subject(:service) { described_class.new(goods_nomenclature_item_id).call }
```
See
```
spec/services/applicable_vat_options_service_spec.rb|2 col 3| subject(:service) { described_class.new(measures) }
spec/services/cached_geographical_area_service_spec.rb|2 col 3| subject(:service) { described_class.new(actual_date, countries) }
spec/services/cached_commodity_service_spec.rb|2 col 3| subject(:service) { described_class.new(commodity.reload, actual_date, filter_params) }
spec/services/quota_search_service_spec.rb|2 col 3| subject(:service) { described_class.new(filter, current_page, per_page) }
spec/services/applicable_additional_code_service_spec.rb|2 col 3| subject(:service) { described_class.new(measures) }
spec/services/exchange_rate_service_spec.rb|2 col 3| subject(:service) { described_class.new }
spec/services/measure_unit_service_spec.rb|2 col 3| subject(:service) { described_class.new(measures) }
```

File: spec/services/find_ancestors_service_spec.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+  context 'when the search term is empty' do
+    let(:goods_nomenclature_item_id) { '' }
+
+    it 'returns no ancestors' do
+      find_ancestors
```
Comment (willfish) [inline spec/services/find_ancestors_service_spec.rb:8]: I don't think you need this line?

File: spec/services/find_ancestors_service_spec.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+      expect(find_ancestors).to eq([])
+    end
+  end
+
+  context 'when the search term is Fully qualified (10 chars)' do
```
Comment (willfish) [inline spec/services/find_ancestors_service_spec.rb:14]:
```suggestion
  context 'when the search term is fully qualified (10 chars)' do
```

File: spec/services/find_ancestors_service_spec.rb [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    let(:commodity) { create :commodity }
+    let(:goods_nomenclature_item_id) { commodity.goods_nomenclature_item_id }
+
+    it 'returns the correct ancestors' do
+      find_ancestors
```
Comment (willfish) [inline spec/services/find_ancestors_service_spec.rb:19]: Do you need the service to execute before you run specify the expected result?

Comment (willfish) [reply spec/services/find_ancestors_service_spec.rb:19]: Okay, I can see that you do

File: spec/services/find_ancestors_service_spec.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+    it 'returns no ancestors' do
+      find_ancestors
+
+      expect(find_ancestors).to eq([])
+    end
+  end
+
+  context 'when the search term is Fully qualified (10 chars)' do
+    let(:commodity) { create :commodity }
```
Comment (willfish) [inline spec/services/find_ancestors_service_spec.rb:15]: How do you feel about not mixing dsl/non-dsl styles for method invocations? Happy if you both prefer this, though

```suggestion
    let(:commodity) { create(:commodity) }
```

File: spec/services/quota_search_service_spec.rb [hunk 10]
```diff
@@ -52,14 +53,22 @@
   end
 
   describe '#call' do
-    context 'when filtering by the goods_nomenclature_item_id' do
+    context 'when filtering by a fully qualified goods_nomenclature_item_id' do
```
Comment (willfish) [inline spec/services/quota_search_service_spec.rb:56]:
```suggestion
    context 'when filtering by a fully-qualified goods_nomenclature_item_id' do
```

File: spec/services/quota_search_service_spec.rb [hunk 11]
```diff
@@ -52,14 +53,22 @@
...
       let(:filter) { { 'goods_nomenclature_item_id' => measure1.goods_nomenclature_item_id } }
 
       it 'returns the correct quota definition' do
         expect(service.call).to eq([quota_definition1])
       end
     end
 
+    context 'when filtering by a NOT fully qualified goods_nomenclature_item_id' do
```
Comment (willfish) [inline spec/services/quota_search_service_spec.rb:64]:
```suggestion
    context 'when filtering by a NOT fully-qualified goods_nomenclature_item_id' do
```

File: spec/services/quota_search_service_spec.rb [hunk 12]
```diff
@@ -87,7 +96,7 @@
     context 'when filtering by status exhausted' do
       let(:filter) { { 'status' => 'exhausted' } }
 
-      let!(:quota_exhaustion_event) do
+      before do
```
Comment (willfish) [inline spec/services/quota_search_service_spec.rb:99]: Love it!

**Comment by willfish** [comment]
Re N + 1s @jebw  I believe the ancestors method is already immune to that (it pulls out indent and description relationships with an eager load.

File: spec/services/quota_search_service_spec.rb (hunk 2 above)
Comment (willfish) [reply spec/services/quota_search_service_spec.rb:1]: Maybe do that in a separate PR @alexdesi and we can just accept that the linting doesn't pass for this file for now

Comment (alexdesi) [reply spec/services/quota_search_service_spec.rb:1]: Since it is not a fast thing, yes it's better to keep it disabled for now, so Madhu can start testing this on Monday

File: app/services/find_ancestors_service.rb (hunk 3 above)
Comment (alexdesi) [reply app/services/find_ancestors_service.rb]: no, it is considered unsafe by brakman

File: lib/string_to_class.rb (hunk 4 above)
Comment (alexdesi) [reply lib/string_to_class.rb]: that's because brakeman do not like it: see
https://stackoverflow.com/questions/23741259/brakeman-unsafe-reflection-method-constantize-called-with-model-attribute

anyway, 
I removed the module, and just using just a method now

The alternative would be to disable that check, but I guess it's better to stay safe and follow brakeman.

File: spec/services/find_ancestors_service_spec.rb (hunk 5 above)
Comment (alexdesi) [reply spec/services/find_ancestors_service_spec.rb:2]: ok I'll change it,
but what do you think it's more readable
`find_ancestor` 
or
`service`?
😺

File: spec/services/find_ancestors_service_spec.rb (hunk 6 above)
Comment (alexdesi) [reply spec/services/find_ancestors_service_spec.rb:8]: good catch 👍

File: app/services/find_ancestors_service.rb (hunk 3 above)
Comment (willfish) [reply app/services/find_ancestors_service.rb]: That's a bit nuts, really. It is safe

File: lib/string_to_class.rb (hunk 4 above)
Comment (willfish) [reply lib/string_to_class.rb]: I think this should move to the class_determinator

File: spec/services/find_ancestors_service_spec.rb (hunk 5 above)
Comment (willfish) [reply spec/services/find_ancestors_service_spec.rb:2]: When it comes to the subject under test I normally just look at the subject description (class name) at the top of the file. It's a bit like using `described_class`  the fully qualified name versus `FindAncestors`. I guess the service pattern is so common it makes sense to just follow a standard that codes to the interface and not to this specific implementation.

File: app/models/goods_nomenclature.rb [hunk 13]
```diff
@@ -128,4 +128,15 @@ def code
   def bti_url
     'https://www.gov.uk/guidance/check-what-youll-need-to-get-a-legally-binding-decision-on-a-commodity-code'
   end
+
+  # This method is safer than `constantize`
+  # which is considered dangeros by `brakeman` (a static analyzer)
+  def self.constantize_class_name(klass_name)
```
Comment (willfish) [inline app/models/goods_nomenclature.rb:134]: We can just call `GoodsNomenclature.sti_load(klass_name)` and remove this

Comment (willfish) [inline app/models/goods_nomenclature.rb:134]: This feels really wrong :thinking:

Comment (willfish) [reply app/models/goods_nomenclature.rb:134]:
```ruby
GoodsNomenclature.sti_load(goods_nomenclature_item_id: '1234000000').class => Heading
```

Comment (willfish) [reply app/models/goods_nomenclature.rb:134]: I think we can get rid of this and just use the methods in the Sti plugin

**Review by willfish — APPROVED** [review]
